plutus-tx-plugin/test/Budget/9.6/builtinListIndexing.pir.golden

```diff
@@ -0,0 +1,32 @@
+let
+  data Unit | Unit_match where
+    Unit : Unit
+in
+letrec
+  !go : list data -> integer -> data
+    = \(xs : list data) (i : integer) ->
+        chooseList
+          {data}
+          {Unit -> Unit -> data}
+          xs
+          (\(ds : Unit) (eta : Unit) -> error {data})
```

Comment:
Ohhhhhh...

So we clearly don't need two `Unit` arguments, how do we get them? Well, you wrote

```haskell
    go :: BI.BuiltinList a -> Integer -> a
    go xs i =
      Builtins.matchList
        xs
        (\_ -> traceError builtinListIndexTooLargeError)
        ( \hd tl _ ->
            if i `Builtins.equalsInteger` 0
              then hd
              else go tl (Builtins.subtractInteger i 1)
        )
        ()
```

which makes sense, you don't want to evaluate the nil case and fail at the very first element.

But why two `Unit`? That's because `matchList` has a `Unit` inside of it to make pattern matching on lists lazy:

```haskell
matchList :: forall a r . BI.BuiltinList a -> r -> (a -> BI.BuiltinList a -> r) -> r
matchList l nilCase consCase = BI.chooseList l (const nilCase) (\_ -> consCase (BI.head l) (BI.tail l)) ()
```

But it doesn't make the matching actually lazy! `matchList` always forces `nilCase`, so putting `()` inside of it doesn't help with laziness as that point `nilCase` is already forced. So you have to do the `()` dance again and end up with two unit arguments, which is kinda ridiculous.

I think we should have two `matchList` functions instead:

```haskell
matchList :: forall a r . BI.BuiltinList a -> (() -> r) -> (a -> BI.BuiltinList a -> r) -> r
matchList' :: forall a r . BI.BuiltinList a -> r -> (a -> BI.BuiltinList a -> r) -> r
```

one for lazy matching and one for strict matching.

And while we're here we should probably have some kind of `Lazy` data type instead of `() ->`, so that we can compile it to `delay` and `force` instead of juggling the slower units.

Reply:
Maybe this is why we got such a dramatic reduction in MEM in your [earlier](https://github.com/IntersectMBO/plutus/pull/5896) PR by the way? Although I just took a cursory look and was unable to spot anything.

Reply:
That doesn't solve the "two unit arguments" problem, does it? You'd still end up with `matchList l nilCase consCase = BI.chooseList l (const nilCase) ...`, in other words `BI.chooseList` has to add an additional unit.

Reply:
No, I think it should solve the problem for the most part. You'll end up with

```
matchList l nilCase consCase = BI.chooseList l nilCase (const <...>) ()
```

Note the absence of `const` in the `nilCase` case. It'll force `nilCase` to WHNF, but that's a lambda and so you won't actually evaluate the result, just reach the lambda. We don't need that for `consCase`, since it already evaluates to a lambda, being a function.

This doesn't solve all problems, in particular if `consCase` is fully eta-reduced, then you'll end up evaluating both branches, but that's the usual caveat of strict languages and the answer is the usual too: don't fully eta-reduce your functions unless you actually want to evaluate the eta-reduced function.

tl;dr you'll keep the unit in the definition of `matchList` but won't add it to the call site of `matchList`.

Reply:
I agree that `matchList` is silly and shouldn't be like that. I'm not sure we need two? We just need to delete the pointless units.

Reply:
You are right - I removed the redundant unit and opened https://github.com/IntersectMBO/plutus/issues/5908 for further optimization.

Reply:
I was looking at this and I realised the discussion here is slightly wrong. The delays in `matchList` are not about the _nil_ case, they're about the _cons_ case. In the cons case we call `head` and `tail`; we don't want to do that unless we're definitely choosing that case (they'll fail!). So we have to delay the cons case, and that means we have to delay the nil case too.

Reply:
> I was looking at this and I realised the discussion here is slightly wrong. The delays in matchList are not about the nil case, they're about the cons case. 

The nil case is equally important, since it's an arbitrary expression and you may as well have some `head` in there as well. Except

> In the cons case we call `head` and `tail`; we don't want to do that unless we're definitely choosing that case (they'll fail!). 

in the cons case in `matchList` (as opposed to `chooseList`) we have a function, i.e. we don't call `head` and `tail` unless the function is actually called.

So it _is_ mostly about the nil case, not the cons case in `matchList`. In `chooseList` both are equally important.